packages/connect-agreement/src/models/Agreement.ts

```diff
@@ -147,7 +153,69 @@ export default class Agreement {
     callback?: SubscriptionCallback<StakingMovement[]>
   ): SubscriptionResult<StakingMovement[]> {
     return subscription<StakingMovement[]>(callback, (callback) =>
-      this.#connector.onStakingMovements(this.stakingId(tokenAddress, userAddress), this.#address, first, skip, callback)
+      this.#connector.onStakingMovements(this.stakingId(tokenAddress, userAddress), this.address, first, skip, callback)
     )
   }
+
+  actionId(actionNumber: string): string {
+    return `${this.address.toLowerCase()}-action-${actionNumber}`
+  }
+
+  async action(actionNumber: string): Promise<Action | null> {
+    return this.#connector.action(this.actionId(actionNumber))
+  }
+
+  onAction(acitonNumber: string, callback?: SubscriptionCallback<Action | null>): SubscriptionResult<Action | null> {
+    return subscription<Action | null>(callback, (callback) =>
+      this.#connector.onAction(this.actionId(acitonNumber), callback)
+    )
+  }
+
+  sign(signerAddress: string, versionNumber: string): Promise<ForwardingPath> {
+    return this.#app.intent('sign', [versionNumber], { actAs: signerAddress })
+  }
+
+  async challenge(actionNumber: string, settlementOffer: string, finishedEvidence: boolean, context: string, signerAddress: string): Promise<ForwardingPath> {
+    const intent = await this.#app.intent('challengeAction', [actionNumber, settlementOffer, finishedEvidence, utils.toUtf8Bytes(context)], { actAs: signerAddress })
+
+    const action = (await this.action(actionNumber))!
```

Comment:
Instead of using `!`: should we test for it to be `null`, throw an error if that’s the case, then use `as Action`?